internal/runnerinstall/docker.go

```diff
@@ -0,0 +1,198 @@
+package runnerinstall
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/hashicorp/waypoint/internal/pkg/flag"
+)
+
+type DockerConfig struct {
+	RunnerImage string `hcl:"runner_image,optional"`
+	Network     string `hcl:"network,optional"`
+}
+
+type DockerRunnerInstaller struct {
+	Config DockerConfig
+}
+
+func (i *DockerRunnerInstaller) Install(ctx context.Context, opts *InstallOpts) error {
+	ui := opts.UI
+	sg := opts.UI.StepGroup()
+	defer sg.Wait()
+
+	s := sg.Add("Initializing Docker client...")
+	defer func() { s.Abort() }()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return err
+	}
+	cli.NegotiateAPIVersion(ctx)
+
+	runnerImage := i.Config.RunnerImage
+	imageRef, err := reference.ParseNormalizedNamed(runnerImage)
+
+	imageList, err := cli.ImageList(ctx, types.ImageListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "reference",
+			Value: reference.FamiliarString(imageRef),
+		}),
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(imageList) == 0 {
+		s.Update("Pulling image %s", runnerImage)
+
+		resp, err := cli.ImagePull(ctx, reference.FamiliarName(imageRef), types.ImagePullOptions{})
+		if err != nil {
+			s.Update("Unable to pull waypoint image")
+			return err
+		}
+		defer resp.Close()
+
+		stdout, _, err := ui.OutputWriters()
+		if err != nil {
+			return err
+		}
+
+		var termFd uintptr
+		if f, ok := stdout.(*os.File); ok {
+			termFd = f.Fd()
+		}
+
+		err = jsonmessage.DisplayJSONMessagesStream(resp, s.TermOutput(), termFd, true, nil)
+		if err != nil {
+			return fmt.Errorf("unable to stream pull logs to the terminal: %s", err)
+		}
+
+	}
+
+	var waypointNetwork network.NetworkingConfig
+	if i.Config.Network != "" {
+		waypointNetwork = network.NetworkingConfig{
+			EndpointsConfig: map[string]*network.EndpointSettings{
+				i.Config.Network: {},
+			},
+		}
+	}
```

Comment:
I do remember this being weird! This is because the network name (provided by the user) is the in the key portion of the map element, and the value portion of the map element is an empty struct because we have no opinion on the `EndpointSettings` for the network. Will definitely add a comment though! 😄